Source/SwiftLintBuiltInRules/Rules/Style/PreferSelfInStaticReferencesRule.swift

```diff
@@ -180,6 +180,11 @@ private class Visitor: ViolationsSyntaxVisitor {
     }
 
     override func visitPost(_ node: SimpleTypeIdentifierSyntax) {
+        if let grandParent = node.parent?.parent,
+           case .likeClass = parentDeclScopes.peek(), grandParent.is(GenericArgumentSyntax.self) {
+            // Type is a generic parameter in a class.
+            return
+        }
```

Comment:
What about this (admittedly rather pathological) example?

```swift
class C1<T> {}
class C2: C1<C2.C3.C4> {
    class C3 {
        class C4 {}
    }
}
```

I assume it would fail the test.

It's easier to skip `GenericArgumentListSyntax` elements entirely when we're in a class scope. That can be done by another `visit` method. Looking up parents is only useful for specific cases. When recursion comes into play this typically fails.

Reply:
Yeah, I don't think recursion makes sense here. I just removed checking `GenericArgumentListSyntax` as suggested and tests now pass. Just couldn't understand this part:

> That can be done by another `visit` method

Reply:
With that change, the rule doesn't trigger on cases like [this](https://github.com/JohnCoates/Aerial/blob/f5a42f790e7e09e9461c2faf0f1b79ea9c98c67e/Aerial/Source/Views/Layers/AnimationLayer.swift#L35:35) anymore. We might need another test case for that. 😅

This is the reason why I suggested:

> That can be done by another `visit` method.

Without verifying it myself, I think adding the method

```swift
override func visit(_ node: GenericArgumentListSyntax) -> SyntaxVisitorContinueKind {
    if case .likeClass = parentDeclScopes.peek() {
        return .skipChildren
    }
    return .visitChildren
}
```

would resolve the conflict.

Reply:
You basically resolved it yourself 😄. Pushed the changes, it worked locally, waiting for CI to finish.

Reply:
No worries. That's called _knowledge sharing_. 😉